### PR TITLE
DKG - Switch to V1

### PR DESCRIPTION
**Pull request by benr-ml**
## Description 

Refactor the DKG to allow protocol upgrades of the cryptographic protocol, and migrate to V1 (controlled by a protocol config parameter).

## Test plan 

Existing tests and the "Continuous" GH workflow.
 
---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [x] Protocol: 
- [x] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (aschran) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:508]: following the pattern for other fields where we've added versionining to messages that are part of these tables, these should end in a _v2 suffix rather than having "versioned" in the name

(I'm not sure I actually like that better but it seems best to stick with current conveniton rather than adding a second one?)

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (aschran) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:3418]: a lot of this duplicated code could be factored out, right? since the only thing that's different is the deserialization call, but the error handling and the call to `add_message` are the same

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (aschran) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:3466]: same here re: factoring

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:69]: maybe simpler to assert_eq! for these?

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:96]: rather than having these if let-else panics in the business logic code path, the pattern we've followed in this case is to add methods for `unwrap_v0` and `unwrap_v1`.

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:340]: doesn't this still need to load from the old tables if the protocol config has dkg version set to 0?

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:452]: again wondering can this be a method on VersionedUsedProcessedMessages? just the part that farms out the call to `complete` or `complete_v1` based on the input message type

**Comment by aschran** [comment]
@mystenmark adding you as backup in case I go OOO before this is ready to merge

File: crates/sui-core/src/epoch/randomness.rs
Comment (benr-ml) [reply crates/sui-core/src/epoch/randomness.rs:340]: once we run the new binary, we use the new tables for both versions.
that said, there could be a problem if the binary is replaced during the dkg, as some data is stored in the old tables while after binary upgrade, the validator is using the new ones.
added code to load from the old tables in that case as well.

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:370]: This should work just based on the protocol version and not always checking both tables, same as we do it for e.g. `pending_checkpoints_v2` table in `AuthorityPerEpochStore`

then once new protocol version is in use everywhere we can clean up the branch

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:419]: I don't think so? it's the responsibility of commit handler to call that. do you see a scenario where it could fail to do so?

**Review by aschran — COMMENTED** [review]
thanks, looking pretty good, adding last nit and an answer to your todo

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (aschran) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:509]: i'd not mark these as deprecated until we actually remove the code that uses the old tables

File: crates/sui-core/src/epoch/randomness.rs
Comment (aschran) [inline crates/sui-core/src/epoch/randomness.rs:520]: i think for consistency this should be based on protocol version and using unwrap_v0 and unwrap_v1, rather than deciding on which table to write to based on the message version

same comment applies below as well

**Review by aschran — APPROVED** [review]
thanks for all the updates, just have a few final nits and clicking approve so you can merge after that

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (benr-ml) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:509]: i'd rather be explicit about it, they should not be used for new code

**Pull request merged**
